pennylane/hf/dipole.py

```diff
@@ -0,0 +1,401 @@
+# Copyright 2018-2022 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+This module contains the functions needed for computing the dipole moment.
+"""
+import autograd.numpy as anp
+import pennylane as qml
+from pennylane import numpy as np
+from pennylane.hf.basis_data import atomic_numbers
+from pennylane.hf.hamiltonian import _generate_qubit_operator, _return_pauli, simplify
+from pennylane.hf.matrices import moment_matrix
+
+
+def dipole_integrals(mol, core=None, active=None):
+    r"""Return a function that computes the dipole moment integrals over the molecular orbitals.
+
+    These integrals are required to construct the dipole operator in the second-quantized form
+
+    .. math::
+
+        \hat{D} = -\sum_{pq} d_{pq} [\hat{c}_{p\uparrow}^\dagger \hat{c}_{q\uparrow} +
+        \hat{c}_{p\downarrow}^\dagger \hat{c}_{q\downarrow}] -
+        \hat{D}_\mathrm{c} + \hat{D}_\mathrm{n},
+
+    where the matrix elements :math:`d_{pq}` are given by the integral over molecular orbitals
+    :math:`\phi`
+
+    .. math::
+
+        d_{pq} = \int \phi_p^*(r) \hat{{\bf r}} \phi_q(r) dr,
+
+    :math:`\hat{c}^{\dagger}` and :math:`\hat{c}` are the creation and annihilation operators,
+    respectively, and :math:`\hat{D}_\mathrm{c}` and :math:`\hat{D}_\mathrm{n}` are the contribution
+    of the core orbitals and nuclei to the dipole operator, respectivel.
+
+    The molecular orbitals are represented as a linear combination of atomic orbitals as
+
+    .. math::
+
+        \phi_i(r) = \sum_{\nu}c_{\nu}^i \chi_{\nu}(r).
+
+    Using this equation the dipole moment integral :math:`d_{pq}` can be written as
+
+    .. math::
+
+        d_{pq} = \sum_{\mu \nu} C_{p \mu} d_{\mu \nu} C_{\nu q},
+
+    where :math:`d_{\mu \nu}` is the dipole moment integral over the atomic orbitals and :math:`C`
+    is the molecular orbital expansion coefficient matrix. The contribution of the core molecular
+    orbitals to the dipole operator is computed as
+
+    .. math::
+
+        \hat{D}_\mathrm{c} = \sum_{i=1}^{N_\mathrm{core}} d_{ii},
```

Comment:
If N_{core} is the number of orbitals, we need a 2 in front of the sum to account for the contribution of both spin projections.
```suggestion
        \hat{D}_\mathrm{c} = 2 \sum_{i=1}^{N_\mathrm{core}} d_{ii},
```